common/formatting/layout_optimizer.h

```diff
@@ -0,0 +1,445 @@
+
+  // Merges sublayouts, horizontally
+  kLayoutHorizontalMerge,
+
+  // Merges sublayout, vertically
+  kLayoutVerticalMerge,
+
+  // Indent sublayout
+  // ----------------
+  // Why do we introduce new layout instead of using
+  // kLayoutHorizontalMerge + kLayoutLine(empty + indent)?
+  // Because:
+  // 1. It does not introduce Knot at column_limit - indent
+  // 2. No need to check layouts type in HorizontalJoin() whether to
+  //    skip or not before_spaces (when merging indent line with normal
+  //    line/layout)
+  // 3. Wrapping layout forces GetSpacesBefore() return 0
+  kLayoutIndent,
+};
+
+std::ostream& operator<<(std::ostream& stream, const LayoutType& type);
+
+// Intermediate partition tree layout
+class Layout {
+ public:
+  Layout(LayoutType type, int spacing)
+      : type_(type), indentation_(0), spaces_before_(spacing) {}
+
+  Layout(const UnwrappedLine& uwline)
+      : type_(LayoutType::kLayoutLine), indentation_(0) {
+    tokens_ = uwline.TokensRange();
+    if (tokens_.size() > 0) {
+      assert(tokens_.size() > 0);
+      spaces_before_ = tokens_.front().before.spaces_required;
+    } else {
+      spaces_before_ = 0;
+    }
+  }
+
+  Layout(int indent)
+      : type_(LayoutType::kLayoutIndent),
+        indentation_(indent),
+        spaces_before_(0) {}
+
+  ~Layout() = default;
+  Layout(const Layout&) = default;
+
+  // Deleting standard interfaces
+  Layout() = delete;
+  Layout(Layout&&) = delete;
+  Layout& operator=(const Layout&) = delete;
+  Layout& operator=(Layout&&) = delete;
+
+  LayoutType GetType() const { return type_; }
+
+  int GetIndentationSpaces() const { return indentation_; }
+
+  int SpacesBeforeLayout() const { return spaces_before_; }
+
+  std::string Text() const {
+    return absl::StrJoin(tokens_, " ",
+                         [=](std::string* out, const PreFormatToken& token) {
+                           absl::StrAppend(out, token.Text());
+                         });
+  }
+
+  UnwrappedLine AsUnwrappedLine() const {
+    assert(type_ == LayoutType::kLayoutLine);
+    UnwrappedLine uwline(0, tokens_.begin());
+    uwline.SpanUpToToken(tokens_.end());
+    uwline.SetPartitionPolicy(PartitionPolicyEnum::kAlwaysExpand);
+    return uwline;
+  }
+
+  int Length() const {
+    assert(type_ == LayoutType::kLayoutLine);
+    if (tokens_.size() == 0) {
+      return 0;
+    }
+    int len = 0;
+    for (const auto& token : tokens_) {
+      len += token.before.spaces_required;
+      len += token.Length();
+    }
+    len -= tokens_.front().before.spaces_required;
+    return len;
+  }
+
+  static int UnwrappedLineLength(const UnwrappedLine& uwline) {
+    int len = 0;
+    for (const auto& token : uwline.TokensRange()) {
+      len += token.before.spaces_required;
+      len += token.Length();
+    }
+    len -= uwline.TokensRange().front().before.spaces_required;
+    return len;
+  }
+
+  SpacingOptions SpacingOptionsLayout() const {
+    assert(type_ == LayoutType::kLayoutLine);
```

Comment:
`CHECK_EQ(type_, LayoutType::kLayoutLine)`, and `CHECK(tokens_.empty());`

Reply:
Done